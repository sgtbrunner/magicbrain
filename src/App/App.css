button {
  cursor: pointer;
}

button:disabled {
  cursor: not-allowed;
}

form {
  background-color: rgba(220, 200, 255, 0.5);
}

.App {
  text-align: center;
}

.smaller {
  transform: scale(0.95);
}

.centered {
  display: flex;
  justify-content: center;
}

.particles {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: -1;
}

.noselect {
  user-select: none;
}
