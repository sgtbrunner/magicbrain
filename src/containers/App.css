.App {
  text-align: center;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

.smaller {
  transform: scale(0.9);
}

.centered {
  display: flex;
  justify-content: center;
}

.form {
background: linear-gradient(to right, rgba(239,239,239,100%) 0%,rgba(219,219,219,100%) 50%,rgba(239,239,239,100%) 100%);
}

.particles {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: -1;
}

.noselect {
  user-select: none;
}
