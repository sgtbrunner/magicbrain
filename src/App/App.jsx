import React, { useState, useEffect } from 'react';
import { Switch, Route, Redirect, useHistory } from 'react-router-dom';
import Particles from 'react-particles-js';

import Header from '../components/Header';
import Home from '../pages/Home';
import NotFound from '../pages/NotFound';
import Register from '../pages/Register';
import SignIn from '../pages/SignIn';
import api from '../utils/api.utils';
import { PARTICLE_OPTIONS } from '../utils/constants.utils';
import { getUserFromLocalStorage } from '../utils/functions.utils';
import 'tachyons';
import './App.css';

const INITIAL_USER_DATA = {
  id: '',
  name: '',
  email: '',
  entries: 0,
  joined: '',
};

const App = () => {
  const [state, setState] = useState({
    input: '',
    imageUrl: '',
  });

  const [user, setUser] = useState(INITIAL_USER_DATA);

  const [box, setBox] = useState({
    topRow: 0,
    rightCol: 0,
    bottomRow: 0,
    leftCol: 0,
  });

  const history = useHistory();
  const { input, imageUrl } = state;
  const { id, name, entries } = user;

  useEffect(() => {
    const loggedInUser = getUserFromLocalStorage();
    if (loggedInUser) {
      const foundUser = JSON.parse(loggedInUser);
      setUser(foundUser);
      history.push('/home');
    }
  }, [history]);

  const loadUser = (data) => {
    setUser({
      id: data.id,
      name: data.name,
      email: data.email,
      entries: Number(data.entries),
      joined: data.joined,
    });
    history.push('/home');
  };

  const onInputChange = (event) => {
    setState({ ...state, input: event.target.value });
  };

  const calculateFaceLocation = (response) => {
    const clarifaiFace = response.outputs[0].data.regions[0].region_info.bounding_box;
    const image = document.getElementById('inputimage');
    const width = Number(image.width);
    const height = Number(image.height);
    return {
      leftCol: clarifaiFace.left_col * width,
      topRow: clarifaiFace.top_row * height,
      rightCol: width - clarifaiFace.right_col * width,
      bottomRow: height - clarifaiFace.bottom_row * height,
    };
  };

  const displayFaceBox = (box) => {
    setBox({ ...box });
  };

  const onImageDetect = () => {
    setState({ ...state, imageUrl: input });
    api
      .getImageBoundary({ input })
      .then((response) => {
        if (response.outputs) {
          api
            .updateImageCount({ id })
            .then((count) => setUser({ ...user, entries: Number(count) }));
        }
        displayFaceBox(calculateFaceLocation(response));
      })
      .catch((err) => {
        alert(err);
        window.alert('Unable to process your image. Please try a different image later!');
      });
  };

  return (
    <div className="App">
      <Particles className="particles" params={PARTICLE_OPTIONS} />
      <Header isUserSignedIn={!!user.id} />
      <Switch>
        <Route exact path="/">
          <Redirect to={user.id ? '/home' : '/signin'} />
        </Route>
        <Route path="/signin" render={() => <SignIn loadUser={loadUser} />} />
        <Route path="/register" render={() => <Register loadUser={loadUser} />} />
        <Route
          path="/home"
          render={() => (
            <Home
              name={name}
              entries={entries}
              imageUrl={imageUrl}
              box={box}
              onInputChange={onInputChange}
              onImageDetect={onImageDetect}
            />
          )}
        />
        <Route component={NotFound} />
      </Switch>
    </div>
  );
};

export default App;
