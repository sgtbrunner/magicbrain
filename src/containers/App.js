import React, {Component} from 'react';
import Particles from 'react-particles-js';
import Navigation from '../components/js/Navigation';
import Rank from '../components/js/Rank';
import Logo from '../components/js/Logo';
import ImageLinkForm from '../components/js/ImageLinkForm';
import FaceRecognition from '../components/js/FaceRecognition';
import SignIn from '../components/js/SignIn';
import Register from '../components/js/Register';
import Clarifai from 'clarifai';
import 'tachyons';
import './App.css';

const app = new Clarifai.App({
  apiKey: 'a9f14c51e25149479e18ee68eb3fa84b'
});

const particlesOptions = {
  particles: {
    number: {
      value:50,
      density: {
        enable:true,
        value_area:500,    
      }
    }
  }
}

class App extends Component {
  // STATE DECLARATION AND INITIALIZATION
  constructor() {
    super()
      this.state = {
        input: '',
        imageUrl: '',
        box: {},
        route: 'signin',
        signedIn: false
    }
  }

  // TRIGGERED EVERYTIME INPUT CHANGES
  onInputChange = (event) => {
    // ** Set input to user value
      this.setState({input: event.target.value})
  }

  // Function used to calculate boundig box boundaries
  calculateFaceLocation = response => {
    const clarifaiFace = response.outputs[0].data.regions[0].region_info.bounding_box;
    const image = document.getElementById('inputimage');
    const width = Number(image.width);
    const height = Number(image.height);
    console.log(width, height);
    return ({
        leftCol: clarifaiFace.left_col * width,
        topRow: clarifaiFace.top_row * height,
        rightCol: width - (clarifaiFace.right_col * width),
        bottomRow: height - (clarifaiFace.bottom_row * height)
    });
  }

  displayFaceBox = box => {
    console.log(box);
    this.setState({box: box});
  }

  // TRIGGERED EVERYTIME BUTTON IS CLICKED
  // app.models.predit is an asynchronous Clarifai syntax
  onButtonClick = () => {
    this.setState({imageUrl: this.state.input});
    app.models.predict(Clarifai.FACE_DETECT_MODEL, this.state.input)
    .then(response => this.displayFaceBox(this.calculateFaceLocation(response)))
    .catch(err => console.log(err)); 
  }

  onRouteChange = (route) => {
    route==='home' ? this.setState({signedIn: true}) : this.setState({signedIn: false});
    this.setState({route: route});
  }

  render() {
  	return (
  	  	<div className="App">
  	  	    <Particles className='particles'
            		params={particlesOptions}
          	/>
  		    <Navigation onRouteChange={this.onRouteChange}
                      signedIn={this.state.signedIn}
          />
          { this.state.route === 'home' ?
              <div>
                <Rank />
                <Logo />
                <ImageLinkForm onInputChange={this.onInputChange} 
                                onButtonClick={this.onButtonClick}
                />
                <FaceRecognition imageUrl={this.state.imageUrl}
                                   box={this.state.box}
                />
              </div>  
    		      : this.state.route === 'signin' ?
                <SignIn onRouteChange={this.onRouteChange}/>
              : 
                <Register onRouteChange={this.onRouteChange}/>
          }
    		</div>            
    );
  }
}

export default App;
