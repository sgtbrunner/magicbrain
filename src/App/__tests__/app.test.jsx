import React from 'react';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';
import 'jest-canvas-mock';

import App from '../App';

const REGISTER = 'Register';
const SIGN_IN = 'Sign In';
const USER_FORM = 'user-form';

const mountApp = () =>
  render(
    <Router>
      <App />
    </Router>
  );

describe('App', () => {
  it.skip('should match snapshot', () => {
    const { container } = mountApp();

    expect(container).toMatchSnapshot();
  });

  it(`should render 'Sign In' and 'Sign Out' link buttons on page header if user
    IS NOT logged in`, () => {
    mountApp();

    const signInButton = screen.getByRole('button', { name: SIGN_IN });
    expect(signInButton).toBeInTheDocument();

    const registerButton = screen.getByRole('button', { name: REGISTER });
    expect(registerButton).toBeInTheDocument();
  });

  it(`should render 'Sign In' form on page if user IS NOT logged in`, () => {
    mountApp();

    const userForm = screen.getByTestId(USER_FORM);
    expect(userForm).toBeInTheDocument();
    expect(userForm).toHaveTextContent('Sign in');
  });

  it(`should render 'Register' form on page if 'Register' link button is 
    clicked and 'Sign In' form again if 'Sign In' link button is clicked, 
    if user IS NOT logged in`, () => {
    mountApp();

    const signInButton = screen.getByRole('button', { name: SIGN_IN });
    const registerButton = screen.getByRole('button', { name: REGISTER });

    userEvent.click(registerButton);

    const registerForm = screen.getByTestId(USER_FORM);
    expect(registerForm).toBeInTheDocument();
    expect(registerForm).toHaveTextContent(REGISTER);

    userEvent.click(signInButton);

    const signInForm = screen.getByTestId(USER_FORM);
    expect(signInForm).toBeInTheDocument();
    expect(signInForm).toHaveTextContent('Sign in');
  });
});
