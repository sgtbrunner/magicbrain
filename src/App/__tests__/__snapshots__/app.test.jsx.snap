// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="particles"
    >
      <canvas
        style="width: 100%; height: 100%;"
      />
    </div>
    <nav
      class="flex justify-end"
    >
      <div>
        <button
          class="f4 link black pv2 ma3 b--none bg-transparent"
          type="button"
        >
          Sign In
        </button>
        <button
          class="f4 link black pv2 ma3 b--none bg-transparent"
          type="button"
        >
          Register
        </button>
      </div>
    </nav>
    <form
      class="br3 ba b--black-10 mv4 mw6 shadow-5 center smaller"
      data-testid="user-form"
    >
      <main
        class="pv3 ph4 black-80"
      >
        <fieldset
          class="ba b--transparent ph0 mh0"
          id="sign-in"
        >
          <h2
            class="f2 ph0 ma0 noselect"
          >
            Sign in
          </h2>
          <div
            class=""
            data-testid="form-input"
          >
            <label
              class="db fw6 lh-copy mv2"
              for="email"
            >
              Email
              <input
                class="pa2 mb0 ba bg-white w-100 br2 () => showError ? 'b--red' : ''"
                id="email"
                name="email"
                type="email"
              />
            </label>
          </div>
          <div
            class="mt4"
            data-testid="form-input"
          >
            <label
              class="db fw6 lh-copy mv2"
              for="password"
            >
              Password
              <input
                class="pa2 mb0 ba bg-white w-100 br2 () => showError ? 'b--red' : ''"
                id="password"
                name="password"
                type="password"
              />
            </label>
          </div>
        </fieldset>
        <div
          class="flex flex-column items-center"
        >
          <button
            class="ph3 pv2 mt3 ba b--black f6 dib "
            disabled=""
            type="submit"
          >
            Sign in
          </button>
          <button
            class="f5 dim black lh-copy mt3 link b--none bg-transparent"
            type="button"
          >
            Register now, it's free!
          </button>
        </div>
      </main>
    </form>
  </div>
</div>
`;
